pkg/global/config.go

```diff
@@ -0,0 +1,81 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package global
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+type ValueType interface {
+	string | int | bool | time.Duration
+}
+
+var configMap sync.Map
+
+func Init(fromConfigMaps ...*corev1.ConfigMap) {
+	configMap.Range(func(key, value interface{}) bool {
+		configMap.Delete(key)
+		return true
+	})
+
+	for _, cm := range fromConfigMaps {
+		if cm == nil {
+			continue
+		}
+
+		for k, v := range cm.Data {
+			configMap.Store(k, v)
+		}
+	}
+}
+
+func Get[T ValueType](name string, defaultValue T) T {
+	v, exists := configMap.Load(name)
+	if !exists {
+		return defaultValue
+	}
+
+	var (
+		returnValue any
+		err         error
+	)
+
+	existing := v.(string)
+
+	switch any(defaultValue).(type) {
+	case string:
+		returnValue = v
+	case int:
+		returnValue, err = strconv.Atoi(existing)
+	case bool:
+		returnValue, err = strconv.ParseBool(existing)
+	case time.Duration:
+		returnValue, err = time.ParseDuration(existing)
+	}
+
+	utilruntime.Must(errors.Wrapf(err, "Failed to parse config option %q", name))
```

Comment:
This creates a denial-of-service risk: since `ConfigMap`s stored in etcd are modifiable by users with sufficient privileges, there is no guarantee that a stored value will have the correct type. Unless I’m missing something, `err` can therefore be non-nil here for reasons other than programmer error.

Instead of panicking, I think the appropriate response here if the value can’t be parsed is to either return an error or log an error and return the default value.

I know the current workqueue config behaviour is to panic on parse errors, but since this is supposed to be more general-purpose I don’t think that’s appropriate.

Reply:
Good point. I'll log an error and return the default.